src/Http/WebUtilities/src/FileBufferingPipeWriter.cs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Internal;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    // TODO: Implement spooling to disk
+    // For perf it would be nice to write straight to the pipe until a threshold, to optimize performance for small payloads.
+    // Later switch to a file stream approach, probably keeping a buffer to write in large chunks to disk.
+    // This approach would be faster for small write but larger memory usage for big ones, so next will require to find new thresholds.
+    // However it would probbaly not meet the buffering requirement.
+    /// <summary>
+    /// A <see cref="Stream"/> that buffers content to be written to disk.
+    /// </summary>
+    public sealed class FileBufferingPipeWriter : PipeWriter, IAsyncDisposable
```

Comment:
I'm not totally sure what the use case for buffering *output* to disk. I can't think of a place where we need that today. I know in Razor's use case we already have a pretty efficient in-memory representation of data, most of the complexity is in avoiding sync over async.

We've also gotten some really significant bad performance feedback about the use of disk for buffering I/O - so we'd like to avoid it when it's not absolutely required.

Reply:
Revisiting this comment, it looks like we do this for XML serialization (which is synch). It's probably fine to leave the XML serialization alone and just let it use streams.

Reply:
This is mainly were I've been stuck for a long time. My understanding is that views can not do chunking, so it requires the whole response to be buffered.
So it is not used only by the XML and Newtonsoft serializers, but also by the view executor: https://github.com/dotnet/aspnetcore/blob/9abf4bfe3f09d0dafc9b7abd8441e6683e9bf612/src/Mvc/Mvc.ViewFeatures/src/ViewComponentResultExecutor.cs#L152

If this is instead only done to avoid sync over async, it will change eveything in this PR.
Or maybe I am missing something on how to avoid chunking.